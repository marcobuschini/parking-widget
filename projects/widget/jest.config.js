module.exports = {
  verbose: true,
  collectCoverage: true,
  reporters: [
    'default',
    [
      'jest-html-reporter',
      {
        pageTitle: 'Test Report',
      },
    ],
  ],
  coverageReporters: ['lcov', 'html'],
}
