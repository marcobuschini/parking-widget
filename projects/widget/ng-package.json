{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/widget",
  "lib": {
    "entryFile": "src/public-api.ts"
  },
  "keepLifecycleScripts": true,
  "assets": ["assets/widget/i18n/*.js"]
}
