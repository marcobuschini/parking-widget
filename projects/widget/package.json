{
  "name": "widget",
  "version": "0.0.1",
  "description": "This is work in progress and it is not supposed to be used in production, test, or development",
  "author": {
    "name" : "Marco Buschini",
    "email" : "marco.buschini@gmail.com",
    "url" : "https://marcobuschini.github.io/"
  },
  "license" : "MIT",
  "peerDependencies": {
    "@angular/animations": "~8.2.14",
    "@angular/cdk": "^8.2.3",
    "@angular/common": "~8.2.14",
    "@angular/compiler": "~8.2.14",
    "@angular/core": "~8.2.14",
    "@angular/flex-layout": "^8.0.0-beta.27",
    "@angular/forms": "~8.2.14",
    "@angular/material": "~8.2.3",
    "@angular/platform-browser": "~8.2.14",
    "@angular/platform-browser-dynamic": "~8.2.14",
    "@angular/router": "~8.2.14",
    "@ngx-translate/core": "^11.0.1",
    "@ngx-translate/http-loader": "^4.0.0",
    "hammerjs": "^2.0.8",
    "rxjs": "^6.5.3",
    "tslib": "^1.10.0",
    "zone.js": "~0.9.1"
  },
  "devDependencies": {
    "@types/jest": "^24.0.23",
    "jest": "^24.9.0",
    "ts-jest": "^24.2.0",
    "jest-html-reporter": "^2.7.0"
  }
}
