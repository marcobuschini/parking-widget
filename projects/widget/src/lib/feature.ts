export class Feature {
    name: string;
}