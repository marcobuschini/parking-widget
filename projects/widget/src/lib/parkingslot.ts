/** A parking slot for sale on the widget. */
export class ParkingSlot {
  /** The unique identifier of the slot. */
  public id: number
  /** The human readable name of the slot. */
  public name: string
  /** An array of the feature names available on the slot. */
  public features: string[]
}
