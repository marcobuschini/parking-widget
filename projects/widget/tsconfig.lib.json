{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "../../out-tsc/lib",
    "target": "es2015",
    "module": "es2015",
    "moduleResolution": "node",
    "declaration": true,
    "sourceMap": true,
    "inlineSources": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "importHelpers": true,
    "types": [],
    "lib": [
      "dom",
      "es2015"
    ]
  },
  "angularCompilerOptions": {
    "enableIvy": false,
    "enableResourceInlining": true,
    "flatModuleId": "widget",
    "flatModuleOutFile": true,
    "fullTemplateTypeCheck": true,
    "skipMetadataEmit": false,
    "skipTemplateCodegen": true,
    "strictMetadataEmit": true,
    "strictInjectionParameters": true
  },
  "exclude": [
    "src/test.ts",
    "**/*.spec.ts"
  ]
}
