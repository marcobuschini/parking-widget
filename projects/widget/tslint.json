{
  "extends": "../../tslint.json",
  "rules": {
    "directive-selector": [true, "attribute", "mlb", "camelCase"],
    "component-selector": [true, "element", "mlb", "kebab-case"]
  }
}
