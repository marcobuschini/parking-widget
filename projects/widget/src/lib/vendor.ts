import { Feature } from './feature';

export class Vendor {
    public name: string;
    public description: string;
    public features: Feature[]
}