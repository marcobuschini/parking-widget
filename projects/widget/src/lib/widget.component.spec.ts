import {
  ComponentFixture,
  TestBed,
  async as realAsync,
  fakeAsync,
  flush,
  flushMicrotasks,
} from '@angular/core/testing'

import { WidgetComponent } from './widget.component'
import { Vendor } from './vendor'
import { HttpClientTestingModule } from '@angular/common/http/testing'
import { of, Observable } from 'rxjs'
import { ParkingSlot } from './parkingslot'
import { Feature } from './feature'
import { MatCardModule } from '@angular/material/card'
import { MatDividerModule } from '@angular/material/divider'
import { MatListModule } from '@angular/material/list'
import { By } from '@angular/platform-browser'

describe('WidgetComponent', () => {
  let component: WidgetComponent
  let fixture: ComponentFixture<WidgetComponent>

  let vendorSpy: jest.SpyInstance<Observable<Feature[]>>
  let slotsSpy: jest.SpyInstance<Observable<ParkingSlot[]>>
  let addSpy: jest.SpyInstance<void, [ParkingSlot?]>

  const dummyVendor: Vendor = {
    name: 'Test Vendor',
    description: 'This is a test vendor',
    features: [
      { name: 'Feature 1' },
      { name: 'Feature 2' },
      { name: 'Feature 3' },
    ],
  }

  const dummySlots: ParkingSlot[] = [
    {
      id: 0,
      name: 'Slot 1',
      features: ['Feature 1', 'Feature 2'],
    },
  ]

  function setTimeoutPromise(milliseconds: number): Promise<void> {
    return new Promise(resolve => {
      setTimeout(resolve, milliseconds)
    })
  }

  beforeEach(async () => {
    TestBed.configureTestingModule({
      declarations: [WidgetComponent],
      imports: [
        MatCardModule,
        MatDividerModule,
        MatListModule,
        HttpClientTestingModule,
      ],
    })
  })

  beforeEach(() => {
    fixture = TestBed.createComponent(WidgetComponent)
    component = fixture.componentInstance
    component.vendor = dummyVendor
    vendorSpy = jest
      .spyOn(component.service, 'getVendorFeatures')
      .mockReturnValue(of(dummyVendor.features))
    slotsSpy = jest
      .spyOn(component.service, 'getParkingSlots')
      .mockReturnValue(of(dummySlots))
    addSpy = jest.spyOn(component.buying, 'emit')

    fixture.detectChanges()
  })

  afterEach(() => {
    component.ngOnDestroy()
  })

  it(
    'should create component with hidden slots, then show, then hide back',
    realAsync(async () => {
      expect(component).toBeTruthy()

      component.ngOnInit()

      await setTimeoutPromise(1000)

      expect(vendorSpy).toHaveBeenCalled()
      expect(slotsSpy).toHaveBeenCalled()

      expect(fixture).toMatchSnapshot()

      let button = fixture.debugElement.queryAll(By.css('button'))[1]
        .nativeElement as HTMLButtonElement
      button.click()
      fixture.detectChanges()
      await setTimeoutPromise(1000)

      expect(fixture).toMatchSnapshot()

      button = fixture.debugElement.queryAll(By.css('button'))[1]
        .nativeElement as HTMLButtonElement
      button.click()
      fixture.detectChanges()
      await setTimeoutPromise(1000)

      expect(fixture).toMatchSnapshot()
    })
  )

  it('should add to cart', () => {
    component.ngOnInit()

    const button = fixture.debugElement.queryAll(By.css('button'))[0]
      .nativeElement as HTMLButtonElement
    button.click()
    fixture.detectChanges()
    expect(addSpy).toHaveBeenCalled()
  })
})
