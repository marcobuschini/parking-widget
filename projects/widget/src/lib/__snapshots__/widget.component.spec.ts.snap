// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetComponent should create component with hidden slots, then show, then hide back 1`] = `
<mlb-parking-widget
  buying={[Function EventEmitter]}
  features={[Function Array]}
  features$={[Function Subscriber]}
  parkingSlots={[Function Array]}
  parkingSlots$={[Function Subscriber]}
  service={[Function WidgetService]}
  showDetails={[Function Array]}
  translate={[Function TranslateService]}
  vendor={[Function Object]}
>
  <mat-card
    class="mat-card"
  >
    <mat-card-title
      class="mat-card-title"
    >
      Test Vendor
    </mat-card-title>
    <mat-card-subtitle
      class="mat-card-subtitle"
    >
      This is a test vendor
    </mat-card-subtitle>
    <mat-card-content
      class="mat-card-content"
    >
      <mat-list
        class="mat-list mat-list-base"
        role="list"
      >
        
        <mat-list-item
          class="mat-list-item"
          role="listitem"
        >
          <div
            class="mat-list-item-content"
          >
            <div
              class="mat-list-item-ripple mat-ripple"
              mat-ripple=""
              ng-reflect-disabled="true"
              ng-reflect-trigger="[object HTMLUnknownElement]"
            />
            <div
              class="mat-list-text"
            />
            Feature 1 
          </div>
        </mat-list-item>
        <mat-list-item
          class="mat-list-item"
          role="listitem"
        >
          <div
            class="mat-list-item-content"
          >
            <div
              class="mat-list-item-ripple mat-ripple"
              mat-ripple=""
              ng-reflect-disabled="true"
              ng-reflect-trigger="[object HTMLUnknownElement]"
            />
            <div
              class="mat-list-text"
            />
            Feature 2 
          </div>
        </mat-list-item>
        <mat-list-item
          class="mat-list-item"
          role="listitem"
        >
          <div
            class="mat-list-item-content"
          >
            <div
              class="mat-list-item-ripple mat-ripple"
              mat-ripple=""
              ng-reflect-disabled="true"
              ng-reflect-trigger="[object HTMLUnknownElement]"
            />
            <div
              class="mat-list-text"
            />
            Feature 3 
          </div>
        </mat-list-item>
      </mat-list>
      <mat-divider
        aria-orientation="horizontal"
        class="mat-divider mat-divider-horizontal"
        role="separator"
      />
      <div
        fxlayout.gt-xs="row"
        fxlayout.lt-sm="column"
        fxlayoutalign="space-between"
      >
        
        <div>
           Slot 1 
          <button
            mat-stroked-button=""
            ng-reflect-translate=""
            translate=""
          >
            Add to cart
          </button>
          
          <button
            mat-stroked-button=""
            ng-reflect-translate=""
            translate=""
          >
            Expand
          </button>
          
          
        </div>
      </div>
    </mat-card-content>
  </mat-card>
</mlb-parking-widget>
`;

exports[`WidgetComponent should create component with hidden slots, then show, then hide back 2`] = `
<mlb-parking-widget
  buying={[Function EventEmitter]}
  features={[Function Array]}
  features$={[Function Subscriber]}
  parkingSlots={[Function Array]}
  parkingSlots$={[Function Subscriber]}
  service={[Function WidgetService]}
  showDetails={[Function Array]}
  translate={[Function TranslateService]}
  vendor={[Function Object]}
>
  <mat-card
    class="mat-card"
  >
    <mat-card-title
      class="mat-card-title"
    >
      Test Vendor
    </mat-card-title>
    <mat-card-subtitle
      class="mat-card-subtitle"
    >
      This is a test vendor
    </mat-card-subtitle>
    <mat-card-content
      class="mat-card-content"
    >
      <mat-list
        class="mat-list mat-list-base"
        role="list"
      >
        
        <mat-list-item
          class="mat-list-item"
          role="listitem"
        >
          <div
            class="mat-list-item-content"
          >
            <div
              class="mat-list-item-ripple mat-ripple"
              mat-ripple=""
              ng-reflect-disabled="true"
              ng-reflect-trigger="[object HTMLUnknownElement]"
            />
            <div
              class="mat-list-text"
            />
            Feature 1 
          </div>
        </mat-list-item>
        <mat-list-item
          class="mat-list-item"
          role="listitem"
        >
          <div
            class="mat-list-item-content"
          >
            <div
              class="mat-list-item-ripple mat-ripple"
              mat-ripple=""
              ng-reflect-disabled="true"
              ng-reflect-trigger="[object HTMLUnknownElement]"
            />
            <div
              class="mat-list-text"
            />
            Feature 2 
          </div>
        </mat-list-item>
        <mat-list-item
          class="mat-list-item"
          role="listitem"
        >
          <div
            class="mat-list-item-content"
          >
            <div
              class="mat-list-item-ripple mat-ripple"
              mat-ripple=""
              ng-reflect-disabled="true"
              ng-reflect-trigger="[object HTMLUnknownElement]"
            />
            <div
              class="mat-list-text"
            />
            Feature 3 
          </div>
        </mat-list-item>
      </mat-list>
      <mat-divider
        aria-orientation="horizontal"
        class="mat-divider mat-divider-horizontal"
        role="separator"
      />
      <div
        fxlayout.gt-xs="row"
        fxlayout.lt-sm="column"
        fxlayoutalign="space-between"
      >
        
        <div>
           Slot 1 
          <button
            mat-stroked-button=""
            ng-reflect-translate=""
            translate=""
          >
            Add to cart
          </button>
          
          
          <button
            mat-stroked-button=""
            ng-reflect-translate=""
            translate=""
          >
            Collapse
          </button>
          
          <mat-list
            class="mat-list mat-list-base"
            role="list"
          >
            
            <mat-list-item
              class="mat-list-item"
              role="listitem"
            >
              <div
                class="mat-list-item-content"
              >
                <div
                  class="mat-list-item-ripple mat-ripple"
                  mat-ripple=""
                  ng-reflect-disabled="true"
                  ng-reflect-trigger="[object HTMLUnknownElement]"
                />
                <div
                  class="mat-list-text"
                />
                Feature 1 
              </div>
            </mat-list-item>
            <mat-list-item
              class="mat-list-item"
              role="listitem"
            >
              <div
                class="mat-list-item-content"
              >
                <div
                  class="mat-list-item-ripple mat-ripple"
                  mat-ripple=""
                  ng-reflect-disabled="true"
                  ng-reflect-trigger="[object HTMLUnknownElement]"
                />
                <div
                  class="mat-list-text"
                />
                Feature 2 
              </div>
            </mat-list-item>
          </mat-list>
        </div>
      </div>
    </mat-card-content>
  </mat-card>
</mlb-parking-widget>
`;

exports[`WidgetComponent should create component with hidden slots, then show, then hide back 3`] = `
<mlb-parking-widget
  buying={[Function EventEmitter]}
  features={[Function Array]}
  features$={[Function Subscriber]}
  parkingSlots={[Function Array]}
  parkingSlots$={[Function Subscriber]}
  service={[Function WidgetService]}
  showDetails={[Function Array]}
  translate={[Function TranslateService]}
  vendor={[Function Object]}
>
  <mat-card
    class="mat-card"
  >
    <mat-card-title
      class="mat-card-title"
    >
      Test Vendor
    </mat-card-title>
    <mat-card-subtitle
      class="mat-card-subtitle"
    >
      This is a test vendor
    </mat-card-subtitle>
    <mat-card-content
      class="mat-card-content"
    >
      <mat-list
        class="mat-list mat-list-base"
        role="list"
      >
        
        <mat-list-item
          class="mat-list-item"
          role="listitem"
        >
          <div
            class="mat-list-item-content"
          >
            <div
              class="mat-list-item-ripple mat-ripple"
              mat-ripple=""
              ng-reflect-disabled="true"
              ng-reflect-trigger="[object HTMLUnknownElement]"
            />
            <div
              class="mat-list-text"
            />
            Feature 1 
          </div>
        </mat-list-item>
        <mat-list-item
          class="mat-list-item"
          role="listitem"
        >
          <div
            class="mat-list-item-content"
          >
            <div
              class="mat-list-item-ripple mat-ripple"
              mat-ripple=""
              ng-reflect-disabled="true"
              ng-reflect-trigger="[object HTMLUnknownElement]"
            />
            <div
              class="mat-list-text"
            />
            Feature 2 
          </div>
        </mat-list-item>
        <mat-list-item
          class="mat-list-item"
          role="listitem"
        >
          <div
            class="mat-list-item-content"
          >
            <div
              class="mat-list-item-ripple mat-ripple"
              mat-ripple=""
              ng-reflect-disabled="true"
              ng-reflect-trigger="[object HTMLUnknownElement]"
            />
            <div
              class="mat-list-text"
            />
            Feature 3 
          </div>
        </mat-list-item>
      </mat-list>
      <mat-divider
        aria-orientation="horizontal"
        class="mat-divider mat-divider-horizontal"
        role="separator"
      />
      <div
        fxlayout.gt-xs="row"
        fxlayout.lt-sm="column"
        fxlayoutalign="space-between"
      >
        
        <div>
           Slot 1 
          <button
            mat-stroked-button=""
            ng-reflect-translate=""
            translate=""
          >
            Add to cart
          </button>
          
          <button
            mat-stroked-button=""
            ng-reflect-translate=""
            translate=""
          >
            Expand
          </button>
          
          
        </div>
      </div>
    </mat-card-content>
  </mat-card>
</mlb-parking-widget>
`;
